### PR TITLE
[SPARK-8057][Core]Call TaskAttemptContext.getTaskAttemptID using Reflection

### PR DESCRIPTION
**Pull request by zsxwing**
Someone may use the Spark core jar in the maven repo with hadoop 1. SPARK-2075 has already resolved the compatibility issue to support it. But `SparkHadoopMapRedUtil.commitTask` broke it recently.

This PR uses Reflection to call `TaskAttemptContext.getTaskAttemptID` to fix the compatibility issue.

### CODE REVIEW

**Comment by zsxwing** [comment]
@pwendell should we get this one into Spark 1.4?

**Comment by srowen** [comment]
Copying the comment here too for continuity that i do not think this is supposed to change. This is solved by compiling Spark for Hadoop 1, as has always been the case, if deploying Spark on Hadoop 1.

**Comment by pwendell** [comment]
I think this affects embedded applications that pull in the spark core jar, and then try to run against a hadoop 2 cluster. Is that right @zsxwing?

**Comment by zsxwing** [comment]
> I think this affects embedded applications that pull in the spark core jar, and then try to run against a hadoop 2 cluster. Is that right @zsxwing?

Right except `hadoop 2` should be `hadoop 1`. Now the spark core jar is compiled with Hadoop 2.

**Comment by srowen** [comment]
Has that ever been a supported use case though? I thought embedding was on-your-own. I think more than just this prevents this from just-working across Hadoop 1 and 2 though right?

**Comment by pwendell** [comment]
There are a large number of applications that embed Spark and they are some of Spark's most active users (e.g. Palantir, etc). If this makes it so that none of them can use Hadoop 1, it seems worth doing. This is an example of something we broke when we changed the compile version to Hadoop 2 for spark-core. Why not just add it if it's so simple?

**Comment by srowen** [comment]
Is this really the only thing that stops it working in this way vs Hadoop 1? I had imagined it was a lot more than this, like transitive dependencies and all that. I can see the practical argument if that's the case, since it's probably infeasible to start publishing twice as many artifacts just for this.

But surely it's not suddenly guaranteed that the artifacts work with both Hadoop 1 and 2 in this way? That's a stronger promise to undertake, and as you say I'm not sure that compatibility has been maintained since for example the Hadoop dependency changed to 2.2. How did it ever continue to work for Hadoop 1 + embedded situations then? maybe it just didn't come up? which is why I wonder if this is all there is.

**Comment by ryan-williams** [comment]
@srowen has convinced me that, rather than try to publish one Spark that works for Hadoops 1 **and** 2, "we" should publish individual artifacts for the Hadoop versions that are not compatible with each other (hopefully just a 1.* and a 2.*).

Conveniently, such artifacts are already built and published at [https://spark.apache.org/downloads.html](https://spark.apache.org/downloads.html), they're just not published anywhere that can be easily programmatically built against, e.g. a Maven repository.

It seems to me that the "correct" solution is to take those already-published artifacts, which people can manually download and run against today, and also publish them to a Maven repository.

Maybe I don't fully understand what is meant by "embedded" Spark, but shouldn't [people that want to "embed" Spark and run against Hadoop 1] simply "embed" one of the Spark JARs that is already built for Hadoop 1 and published and hosted at apache.org? Is it important that they "embed" it via a Maven dependency? 

If so, again, we should publish Maven JARs that are built to support Hadoop 1.

Thanks, let me know if I'm misunderstanding something.

**Comment by srowen** [comment]
If this PR actually makes the current artifact work in embedded contexts for Hadoop 1 as well, then color me surprised but yes I take it all back and think it's worthwhile in practice, theoretical arguments aside. That makes the issue go away much more easily.

Maven artifacts are a quite different thing than the published assemblies at spark.apache.org.

File: core/src/main/scala/org/apache/spark/mapred/SparkHadoopMapRedUtil.scala [hunk 1]
```diff
@@ -22,8 +22,10 @@ import java.lang.reflect.Modifier
 
 import org.apache.hadoop.mapred._
 import org.apache.hadoop.mapreduce.{TaskAttemptContext => MapReduceTaskAttemptContext}
+import org.apache.hadoop.mapreduce.{TaskAttemptID => MapReduceTaskAttemptID}
```
Comment (ryan-williams) [inline core/src/main/scala/org/apache/spark/mapred/SparkHadoopMapRedUtil.scala]: nit: this import isn't used, is it?

**Comment by srowen** [comment]
@zsxwing is that import renaming needed?
Do we have any additional info on whether this is really the only thing stopping the raw Maven artifact from working with Hadoop 1?
It's a fairly minor change and one more bit of reflection won't add too much complexity. If it adds that much value, I think we can merge this.

**Comment by zsxwing** [comment]
> @zsxwing is that import renaming needed?

No. Just removed it.

> Do we have any additional info on whether this is really the only thing stopping the raw Maven artifact from working with Hadoop 1?

I built Core with `Hadoop 1.0.4` and `Hadoop 2.4.0` and verified all classes have the same byte codes except `WriteInputFormatTestDataGenerator.scala` which is only used in python tests.

However, some places about `Parquet` and `Hive` in Spark SQL still have similar issues. I don't know if `Parquet` and `Hive` in Maven are compatible with Hadoop 1. If not, I don't think we should support it in Spark SQL since `Parquet` and `Hive` are already not compatible.

**Comment by zsxwing** [comment]
retest this please

**Comment by shivaram** [comment]
@zsxwing Have you tested this change with a Hadoop1 cluster by any chance ?

**Comment by JoshRosen** [comment]
Is there any easy way to test this locally by configuring the build to use a Hadoop 1.x version?  Or can it only be tested against a real cluster?

**Comment by shivaram** [comment]
I think you can test it locally if you use a package that depends on Spark and hadoop 1.0.4 (with an exclusion for hadoop dependencies being pulled in from Spark -- something like [1]). I couldn't reproduce this with `spark-shell` when I last tried it.

[1] https://github.com/amplab/ml-matrix/blob/master/build.sbt#L23

**Comment by shivaram** [comment]
@zsxwing Any updates on this ? @tomerk just ran into this as well and it'll be good to have a fix for this in 1.5.

**Comment by JoshRosen** [comment]
Just to make sure that this doesn't slip 1.5... if this stands a chance of fixing the issue and doesn't risk the introduction of performance problems or regressions, do you think that we should speculatively merge it now without waiting for a regression test?

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by shivaram** [comment]
Yeah I'd vote for merging this in given its a small change and I don't see it causing further regressions. I can try to test this out sometime, but not sure when I can get to it.

**Comment by shivaram** [comment]
@JoshRosen @zsxwing Any thoughts on merging this for 1.5 ?

**Comment by zsxwing** [comment]
+1 for merging this into 1.5

**Comment by rxin** [comment]
I've merged this.

**Comment by JoshRosen** [comment]
I think that we should also backport this to branch-1.4.

**Comment by zsxwing** [comment]
> I think that we should also backport this to branch-1.4.

+1 since we fix it in 1.5.0. Just confirmed this one didn't have conflicts with branch-1.4.

**Comment by JoshRosen** [comment]
Backported to 1.4.2.

**Pull request closed without merging**